### PR TITLE
TreeDB column store M15A: add protect-only asset reachability

### PR DESCRIPTION
**Pull request by snissn**
## What Landed

M15A adds the production-facing protect-only reachability planner for durable TreeDB column assets, chained on top of #1631.

- Adds `Collection.PlanColumnAssetReachability` for dry-run/protect-only column asset liveness accounting.
- Enumerates active and recovery-authoritative manifest refs through the existing production physical scan snapshot view.
- Accepts explicit candidate, pending-publish, prepared-asset, and pinned-snapshot refs so later M15 chunks can connect durable manager/catalog/pin registries without changing the accounting model.
- Classifies refs as protected, reclaimable, or uncertain, with optional detailed diagnostics.
- Lists isolated typed column asset manager segment files under the column namespace and classifies segments as protected, reclaimable, mixed, unknown, or missing.
- Retains unknown/untracked bytes, missing live segments, invalid refs, zero-byte unknown segments, and cancellation/error uncertainty; M15A performs no deletion, rewrite, or remap.
- Computes rewrite-debt bytes for mixed protected/reclaimable segments while blocking full segment deletion unless the whole segment is reclaimable.

## What Remains Blocked

- M15B: automatic manager/catalog prepared-publish registry integration, snapshot/root pin integration, deletion eligibility, and safe reclaim after roots/pending publishes/snapshots drain.
- M15C: destructive rewrite/remap/compaction cleanup and post-remap validation.
- Full production GC invocation remains protect-only in this PR; no ordinary value-log GC path owns column asset payloads.
- End-to-end destructive lifecycle smoke is intentionally deferred until M15B/M15C can delete/rewrite safely.

## Performance / Benchmark Evidence

Commands run from benchmark head `d169b9dbafee5884d6770cda99e3f42f7a9c49a2` on Apple M3 / darwin arm64. Latest pushed head is `818c30cfa21578c24255eba59b5fef2acf7370f5` with M15A reachability uncertainty/enumeration review hardening plus lower-stack diagnostic naming, actual-worker reporting, report-schema hardening, and fail-closed reason propagation, and preservation of reachability identity on snapshot-view errors.

```sh
go test ./TreeDB/collections -run '^$' -bench 'BenchmarkColumnAssetReachabilityPlanSummaryM15A$' -benchmem -benchtime=3s -count=3
```

| Benchmark | Refs | Segment Bytes | ns/op | MB/sec | B/op | allocs/op |
|---|---:|---:|---:|---:|---:|---:|
| `BenchmarkColumnAssetReachabilityPlanSummaryM15A` | 10,000 | 640,000 | 1,857,255 | 344.59 | 495,131 | 29 |
| `BenchmarkColumnAssetReachabilityPlanSummaryM15A` | 10,000 | 640,000 | 1,728,101 | 370.35 | 495,135 | 29 |
| `BenchmarkColumnAssetReachabilityPlanSummaryM15A` | 10,000 | 640,000 | 1,712,347 | 373.76 | 495,130 | 29 |

Storage/accounting notes for this PR:

- The benchmark is summary-mode planning over 10K `ColumnAssetRef`s in one isolated column asset segment.
- No rows or row value-log payloads are scanned by the planner; the synthetic benchmark input is refs plus segment metadata.
- The planner emits retained/reclaimable/rewrite-debt byte totals; it does not write command-WAL bytes, checkpoint bytes, or DB payload bytes.
- Unknown, non-canonical, zero-byte unknown, or missing-live segment state remains retained/uncertain instead of reclaimable.

## Throughput Interpretation

This is a correctness-first protect-only planner. The relevant gate is that summary planning scales with refs and segment metadata, not row count or full document payloads. The 10K-ref summary benchmark is ~1.7-1.9 ms/op with stable allocation count after cancellation and diagnostic hardening. Detailed entry materialization remains optional, so normal GC eligibility checks do not require detailed ref/segment entries.

## Gate Status

- PASS: `go test ./TreeDB/collections -run 'TestColumnAssetReachability.*M15A' -count=1`
- PASS: `go test ./TreeDB/collections ./cmd/unified_bench -run 'TestColumnAssetReachability.*M15A|TestColumnStoreSuiteThroughputInterpretationM14C|TestColumnStoreSuiteMarkdownRendersThroughputInterpretationM14C|TestColumnStoreSuiteExecutesForcedAggregateAndParallelPhysicalPathsM14B|TestColumnStoreSuitePhysicalPathFailsClosedOnMissingAssetsM14B|TestColumnStoreSuiteQ3ReferenceUsesPhysicalHourSemanticsM14B|TestColumnStoreSuitePhysicalQueryLinesFailsClosedOnUnknownMappingM14B|TestColumnPhysicalQueryParallelMatchesSerialInsertOnlyM14B|TestColumnPhysicalQueryParallelFailsClosedForMutationVisibilityM14B|TestColumnPhysicalScanHonorsCancellationBeforeSchedulingRefsM14B|TestMergeColumnPhysicalQueryDiagnosticsTreatsMutationPartsAsViewLevelM14B|TestColumnQueryPlanner.*M14B|TestColumnQueryPlanner.*M14A|TestColumnPhysical.*M13A|TestColumnManifest.*M13A|TestColumnPhysicalQuery.*M13B|TestColumn.*M13C' -count=1`
- PASS: `go test ./TreeDB/collections ./cmd/unified_bench -count=1`
- PASS: `go test ./TreeDB/collections -run '^$' -bench 'BenchmarkColumnAssetReachabilityPlanSummaryM15A$' -benchmem -benchtime=3s -count=3`
- PASS: `git diff --check`
- PASS: active/recovery manifest refs are protected.
- PASS: candidate-only refs are reclaimable only as dry-run candidates.
- PASS: pending, prepared, and pinned refs override reclaimability and remain protected.
- PASS: duplicate source refs are counted once per source contribution.
- PASS: missing live segments use explicit `missing` segment status.
- PASS: non-canonical and zero-byte unknown segment files fail closed with `Complete=false`.
- PASS: cancellation returns an identity-only incomplete plan instead of misleading partial stats.

## Artifacts

- Final benchmark artifact: `/tmp/gomap_m15a_reachability_bench_solo_20260519_120733.txt`.
- No pprof/profile-dir artifacts were generated for M15A; M15B/M15C destructive eligibility and rewrite/remap work should add richer GC/rewrite profiles.

## Assumption Ledger

- M15A is protect-only. It intentionally cannot delete, rewrite, remap, quarantine, or reclaim physical column assets.
- Candidate, pending, prepared, and pinned refs are explicit API inputs in this PR. M15B/M15C must wire those inputs to durable production registries/root/snapshot state before destructive GC can use the plan.
- Unknown files, missing segments for live refs, invalid refs, cancellation, and untracked bytes are retained/fail-closed by design.
- The planner operates on typed column asset manager refs and isolated segment files, not ordinary `value_vlog` payloads or row documents.

## Summary by CodeRabbit

* **New Features**
  * Added a protect-only dry-run reachability analysis for column assets: classifies refs (protected/reclaimable/uncertain), reports per-segment status (protected/reclaimable/mixed/unknown/missing), handles missing/non-regular/symlink segments, deterministic ordering, optional detailed per-ref/segment output, and computes rewrite-debt bytes.

* **Refactor**
  * Snapshot/scan preparation now respects context cancellation and cleans up acquired snapshots on cancel.

* **Tests**
  * Comprehensive unit tests and a large benchmark covering classifications, segment edge cases, deduplication, cancellation safety, and byte-accounting.

[Review Change Stack](https://app.coderabbit.ai/change-stack/snissn/gomap/pull/1632?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@codex review

Please re-review after propagated M12A pre-WAL declared-column validation fix.

**Comment by snissn** [comment]
@coderabbitai review

File: TreeDB/collections/column_physical_query.go [hunk 1]
```diff
@@ -138,7 +138,7 @@ func (c *Collection) RunColumnPhysicalQueryParallel(req ColumnPhysicalQueryReque
 	}
 	workers := maxWorkers
 	if refs := len(view.AssetRefs); refs <= 1 {
-		return c.RunColumnPhysicalQuery(req)
+		return c.runColumnPhysicalQueryInSnapshotView(view, req)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_physical_query.go]: No code change needed: `RunColumnPhysicalQueryParallel` checks `view.MutationParts > 0` before the single-ref fallback, so mutation-bearing manifests fail closed before `runColumnPhysicalQueryInSnapshotView`. This is covered by the M14B mutation-visibility fail-closed test and the pinned-view single-ref regression.

File: TreeDB/collections/column_physical_query.go [hunk 2]
```diff
@@ -210,6 +210,31 @@ func (c *Collection) RunColumnPhysicalQueryParallel(req ColumnPhysicalQueryReque
...
+		Diagnostics: columnPhysicalQueryDiagnosticsFromScan(diag),
+	}
+	result.Diagnostics.ScanNanos = scanNanos
+	result.Diagnostics.ReduceRows = exec.reduceRows
+	if err != nil {
+		return result, err
+	}
+	result.Groups = exec.groups()
+	result.Diagnostics.ResultGroups = len(result.Groups)
+	return result, nil
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_physical_query.go:243]: No code change needed: `runColumnPhysicalQueryInSnapshotView` is only used after the caller has pinned a snapshot view and rejected mutation-bearing manifests. The parallel entrypoint keeps that fail-closed guard before both multi-ref and single-ref execution.

File: TreeDB/collections/column_asset_reachability.go [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+		status := columnAssetReachabilityStatusForSources(builder.sources)
+		if err := validateColumnAssetRefForPlan(builder.ref); err != nil {
+			status = ColumnAssetReachabilityUncertain
+		}
+		if builder.ref.Namespace != input.namespace {
+			status = ColumnAssetReachabilityUncertain
+		}
+		if builder.ref.Offset > math.MaxInt64-builder.ref.Length {
+			status = ColumnAssetReachabilityUncertain
+		}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_asset_reachability.go:322]: Addressed in the latest #1632 head (`7f8c019b7`): `processRef` now uses the same `columnAssetReachabilityRefCanContributeRange` predicate as range pre-sizing before appending segment ranges, and the invalid-ref regression proves negative offsets do not contribute segment work.

File: TreeDB/collections/column_asset_reachability.go [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+		if status == ColumnAssetReachabilityUncertain {
+			return
+		}
+		rangesByFile[builder.ref.FileID] = append(rangesByFile[builder.ref.FileID], columnAssetReachabilityRange{
+			start:  builder.ref.Offset,
+			end:    builder.ref.Offset + builder.ref.Length,
+			status: status,
+		})
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_asset_reachability.go:353]: Addressed in the latest #1632 head (`7f8c019b7`): invalid refs become uncertain before range append, covering non-positive lengths, zero IDs, wrong namespace, and overflow via the shared range-contribution predicate.

File: TreeDB/collections/column_physical_scan.go [hunk 5]
```diff
@@ -584,8 +587,6 @@ func decodeColumnManifestHeaderRecordForScan(raw []byte) (columnManifestHeaderRe
 	if err := cur.err; err != nil {
 		return columnManifestHeaderRecordForScan{}, err
 	}
-	header.collection = bytes.Clone(header.collection)
-	header.operation = bytes.Clone(header.operation)
 	if header.rowCount > uint64(maxCollectionInt) {
 		return columnManifestHeaderRecordForScan{}, errors.New("collections: column manifest row count overflows int")
 	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_physical_scan.go:632]: Intentional aliasing for the allocation-free planner scan; the propagated 9aa1bab9c comment documents the lifetime contract and the scan does not retain those slices.

File: cmd/unified_bench/suite_column_store_test.go [hunk 6]
```diff
@@ -383,6 +383,22 @@ func TestColumnStoreBenchRunUsesRowsProcessedForPhysicalAggregateM14B(t *testing
...
+		Queries: []columnStoreQueryMetric{
+			{Name: "q1", Rows: 10, RowsPerSecond: 1, duration: 10 * time.Millisecond},
+			{Name: "q2", Rows: 20, RowsPerSecond: 1, duration: 20 * time.Millisecond},
+		},
+	}, nil, 0)
+
+	got := run.Results[columnStoreSuiteBenchTestName][columnStoreSuiteBenchDisplayName]
+	if math.Abs(got-1000) > 1e-9 {
+		t.Fatalf("aggregate rows/sec=%f want 1000 from legacy rows fallback", got)
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply cmd/unified_bench/suite_column_store_test.go:402]: Addressed by the propagated 3ef756075 test fix; the aggregate throughput assertions now use a 1e-6 tolerance.

File: TreeDB/collections/column_query_plan.go [hunk 7]
```diff
@@ -165,7 +170,7 @@ func columnQueryRequestNeedsPhysicalCapabilityDiscovery(req ColumnQueryPlanReque
 	case "":
 		return true
 	default:
-		return false
+		return true
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_query_plan.go]: Intentional fail-closed diagnostic behavior: capability discovery clears caller-supplied stale capability counts and reports manifest-derived diagnostics even for an invalid force kind; it does not schedule physical execution.

File: TreeDB/collections/column_asset_reachability.go [hunk 8]
```diff
@@ -0,0 +1,833 @@
...
+
+func (in *columnAssetReachabilityInput) addRefs(ctx context.Context, refs []ColumnAssetRef, source ColumnAssetReachabilitySource) error {
+	for _, ref := range refs {
+		if ctx != nil && ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if in.addRef(ref, source) {
+			in.incrementSourceCount(source)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_asset_reachability.go]: Addressed in 195d390dc by amortizing reachability context polling and updating cancellation consistency coverage.

**Comment by snissn** [comment]
@coderabbitai review
@codex review

Latest head is `9c1c58dcc4e6a5a0d1e43a1c5ab829d64831aa29`. Please re-review the narrow review-fix follow-up on this chained column-store PR. The stack fixes cover internal-only planner reason strings, manifest capability validation hardening, physical query fail-closed/evidence corrections, amortized reachability cancellation polling, and the M15B GC lock-scope documentation where applicable. Targeted tests passed on the affected branch heads; CI is pending in the GitHub runner backlog.

**Comment by snissn** [comment]
@coderabbitai review\n@codex review\n\nRestacked this chained PR onto its immediate parent branch so GitHub can evaluate the latest head (0eb0af5a30a8a8c59bea5b2fd01bcd67207f341e). The merge commit is stack-shape only; targeted top-of-stack regression tests pass.

**Comment by snissn** [comment]
@coderabbitai review

@codex review

**Comment by snissn** [comment]
@coderabbitai review
@codex review

Latest head includes fixes for the unresolved review comments on this PR/stack. Please re-review against the current head.

**Comment by snissn** [comment]
@coderabbitai review
@codex review

Latest head includes the final reachability/snapshot review-edge fixes. Please re-review against the current head.

**Comment by snissn** [comment]
@coderabbitai review
@codex review

Latest head includes fixes for the newest review comments and has been propagated through the stack. Please re-review against the current head.

**Comment by snissn** [comment]
@coderabbitai review
@codex review

Latest head includes the M14B physical-query review fix (explicit worker index capture and segment-file cache diagnostics), propagated through the stack with tracker/PR body updates. Please re-review against the current head.

**Comment by snissn** [comment]
Latest head pushed and prior review threads resolved. Please re-review this PR in the chained column-store stack. Latest validation includes focused TreeDB/collections and cmd/unified_bench regression tests from the propagated top-of-stack head; no merge requested.

File: TreeDB/collections/column_asset_reachability.go [hunk 9]
```diff
@@ -0,0 +1,896 @@
...
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				continue
+			}
+			return nil, err
+		}
+		if info.IsDir() {
+			continue
+		}
+		fileID, ok := columnAssetReachabilitySegmentFileID(entry.Name())
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_asset_reachability.go]: Fixed in bffc6dc19. Canonical segment-named directories are now retained as zero-byte unknown segment entries instead of being skipped, so the reachability planner fails closed on this corruption/misconfiguration shape. Targeted coverage: `TestColumnAssetReachabilityPlanRetainsCanonicalDirectoryAsUnknownM15A`.

File: TreeDB/collections/column_asset_reachability.go [hunk 10]
```diff
@@ -0,0 +1,896 @@
...
+			continue
+		}
+		if !info.Mode().IsRegular() {
+			if ok {
+				segments = append(segments, columnAssetReachabilitySegment{fileID: fileID, path: path})
+			} else {
+				segments = append(segments, columnAssetReachabilitySegment{path: path})
+			}
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_asset_reachability.go]: Fixed in bffc6dc19. Segment enumeration now constructs the segment entry through a single append helper for directories, symlinks, non-regular files, and regular files, preserving the canonical file id when the name parses and avoiding the duplicated branches.

**Comment by snissn** [comment]
@coderabbitai review latest head after review-thread fixes and stack propagation.

File: TreeDB/collections/column_asset_reachability.go [hunk 11]
```diff
@@ -0,0 +1,884 @@
...
+	case unknownBytes != 0 || outOfBounds != 0:
+		status = ColumnAssetReachabilitySegmentUnknown
+	case protectedBytes != 0 && reclaimableBytes != 0:
+		status = ColumnAssetReachabilitySegmentMixed
+	case protectedBytes != 0:
+		status = ColumnAssetReachabilitySegmentProtected
+	case reclaimableBytes != 0:
+		status = ColumnAssetReachabilitySegmentReclaimable
+	}
+	return columnAssetReachabilitySegmentPlan{
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_asset_reachability.go:602]: Fixed in 10ac0fadf. `classifyColumnAssetReachabilitySegment` now treats a regular canonical segment file with no reachable refs as a complete reclaimable segment, while zero-byte/non-regular canonical entries and non-canonical files remain fail-closed unknown. Regression coverage: `TestColumnAssetReachabilityPlanClassifiesUnreferencedCanonicalSegmentReclaimableM15A`; M15C incomplete-plan coverage was adjusted to use a non-canonical unknown segment.

**Comment by snissn** [comment]
@coderabbitai review latest head after unreferenced segment reachability fix and propagation.

File: TreeDB/collections/column_asset_reachability.go [hunk 12]
```diff
@@ -0,0 +1,884 @@
...
+	}
+	if err := ctx.Err(); err != nil {
+		return ColumnAssetReachabilityPlan{ProtectOnly: true}, err
+	}
+	view, closeView, err := c.prepareColumnPhysicalScanSnapshotViewWithContext(ctx)
+	if closeView != nil {
+		defer closeView()
+	}
+	if err != nil {
+		return ColumnAssetReachabilityPlan{ProtectOnly: true}, err
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_asset_reachability.go]: Fixed in 818c30cfa. PlanColumnAssetReachability now builds the fail-closed identity plan from the partial snapshot view on prepare errors, and TestColumnAssetReachabilityPlanPreservesIdentityOnSnapshotViewErrorM15A covers the incomplete-manifest path without exposing partial stats.

File: TreeDB/collections/column_asset_reachability.go [hunk 13]
```diff
@@ -0,0 +1,894 @@
...
+	plan := columnAssetReachabilityPlanWithStats(input)
+	namespace, err := columnAssetManagerNamespaceForRoot(input.rootDir, input.namespace)
+	if err != nil {
+		plan.Complete = false
+		return plan, err
+	}
+	segments, err := listColumnAssetReachabilitySegments(ctx, namespace.SegmentDir)
+	if err != nil {
+		plan.Complete = false
+		return plan, err
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/collections/column_asset_reachability.go:298]: Current descendant stack already returns the identity/stats plan with Complete=false when segment enumeration is cancelled inside buildColumnAssetReachabilityPlan, rather than an empty successful plan. No destructive GC path treats cancellation as a no-op.
